File: regression.py
"""Shared statistics used across the re-inspection analysis scripts.

analyze_reinspection_model.py and investigate_deeper_mechanisms.py both build
a restaurant-clustered logistic model on the same paired initial -> re-
inspection dataset. This module is the single implementation of that pairing
logic, the hand-rolled IRLS fit, and the cluster-robust standard errors, so a
fix here reaches every script that uses it instead of needing to be repeated
in each one.
"""
from __future__ import annotations

from math import erf, sqrt

import numpy as np
import pandas as pd

INITIAL = "Cycle Inspection / Initial Inspection"
REINSPECTION = "Cycle Inspection / Re-inspection"
BOROUGHS = ["Bronx", "Brooklyn", "Manhattan", "Queens", "Staten Island"]


def normal_p_value(z: float) -> float:
    """Two-sided p-value for a standard-normal z-statistic."""
    return 2 * (1 - 0.5 * (1 + erf(abs(z) / sqrt(2))))


def build_paired_dataset(raw: pd.DataFrame, extra_visit_cols: list[str] | None = None) -> pd.DataFrame:
    """Pair each restaurant's cycle re-inspection with its most recent prior
    cycle-initial inspection. One output row per paired visit, filtered to
    graded (A/B/C) outcomes in the five NYC boroughs from 2016 onward.

    extra_visit_cols: additional fields to carry over from the re-inspection
    row (e.g. "community_board" for a density covariate).
    """
    extra_visit_cols = extra_visit_cols or []
    raw = raw[raw["inspection_type"].isin([INITIAL, REINSPECTION])].copy()
    raw["inspection_date"] = pd.to_datetime(raw["inspection_date"])

    # Each original row is one citation; collapse back to one inspection visit.
    visit = (
        raw.sort_values("inspection_date")
        .groupby(["camis", "inspection_date", "inspection_type"], as_index=False)
        .first()
    )
    visit["score"] = pd.to_numeric(visit["score"], errors="coerce")
    visit["year"] = visit["inspection_date"].dt.year

    paired = []
    for camis, group in visit.groupby("camis", sort=False):
        prior_initial = None
        for row in group.sort_values("inspection_date").itertuples(index=False):
            if row.inspection_type == INITIAL:
                prior_initial = row
            elif row.inspection_type == REINSPECTION and prior_initial is not None:
                record = {
                    "camis": camis,
                    "borough": row.boro,
                    "cuisine": row.cuisine_description,
                    "reinspection_year": row.year,
                    "initial_score": prior_initial.score,
                    "grade": row.grade,
                }
                for col in extra_visit_cols:
                    record[col] = getattr(row, col, None)
                paired.append(record)

    df = pd.DataFrame(paired)
    # Grade A is the reference outcome; retain only completed letter-grade visits.
    df = df[df["grade"].isin(["A", "B", "C"])].dropna(
        subset=["borough", "cuisine", "initial_score", "reinspection_year"]
    )
    df = df[df["borough"].isin(BOROUGHS)]
    df = df[df["reinspection_year"] >= 2016].copy()
    df["b_or_c"] = df["grade"].isin(["B", "C"]).astype(int)
    df["initial_score_10"] = df["initial_score"] / 10.0
    # Too few eligible observations before 2023 to estimate stable
    # calendar-year coefficients; bucket them together instead.
    df["year_model"] = df["reinspection_year"].where(
        df["reinspection_year"] >= 2023, "2022 or earlier"
    ).astype(str)
    # Keep cuisine as a control while avoiding unstable single-observation levels.
    cuisine_count = df["cuisine"].value_counts()
    df["cuisine_model"] = df["cuisine"].where(df["cuisine"].map(cuisine_count) >= 20, "Other / rare cuisine")
    return df


def design_matrix(df: pd.DataFrame, extra_cols: list[str] | None = None) -> tuple[np.ndarray, list[str]]:
    """Borough (Manhattan held out) + cuisine + year dummies, plus
    initial_score_10 and any extra numeric covariates, with an intercept."""
    parts = [df[["initial_score_10"]]]
    if extra_cols:
        parts.append(df[extra_cols])
    parts += [
        pd.get_dummies(df["borough"], prefix="borough", drop_first=False, dtype=float).drop(columns="borough_Manhattan"),
        pd.get_dummies(df["cuisine_model"], prefix="cuisine", drop_first=True, dtype=float),
        pd.get_dummies(df["year_model"], prefix="year", drop_first=True, dtype=float),
    ]
    design = pd.concat(parts, axis=1)
    X = np.column_stack([np.ones(len(df)), design.to_numpy(float)])
    columns = ["Intercept", *design.columns]
    return X, columns


def fit_clustered_logit(X: np.ndarray, y: np.ndarray, clusters: pd.Series):
    """Logistic MLE via iteratively reweighted least squares, with a
    cluster-robust sandwich covariance (clustered on `clusters`, e.g. the
    restaurant ID, so repeated visits from the same restaurant don't count as
    independent observations).

    Returns (beta, se, covariance, n_clusters, n_observations).
    """
    def log_likelihood(coefficients: np.ndarray) -> float:
        linear = np.clip(X @ coefficients, -30, 30)
        return float(np.sum(y * linear - np.logaddexp(0, linear)))

    beta = np.zeros(X.shape[1])
    current_ll = log_likelihood(beta)
    for _ in range(100):
        eta = np.clip(X @ beta, -30, 30)
        mu = 1 / (1 + np.exp(-eta))
        w = np.clip(mu * (1 - mu), 1e-9, None)
        hessian = X.T @ (X * w[:, None])
        score = X.T @ (y - mu)
        step = np.linalg.lstsq(hessian, score, rcond=None)[0]
        # Backtracking keeps IRLS stable when sparse categorical cells create
        # an overly aggressive Newton step.
        multiplier = 1.0
        beta_next = beta + step
        while log_likelihood(beta_next) < current_ll and multiplier > 1e-6:
            multiplier /= 2
            beta_next = beta + multiplier * step
        if np.max(np.abs(multiplier * step)) < 1e-9:
            beta = beta_next
            break
        beta = beta_next
        current_ll = log_likelihood(beta)
    else:
        raise RuntimeError("IRLS did not converge")

    mu = 1 / (1 + np.exp(-np.clip(X @ beta, -30, 30)))
    bread = np.linalg.inv(X.T @ (X * (mu * (1 - mu))[:, None]))
    clusters = pd.Series(clusters).reset_index(drop=True)
    cluster_scores = []
    for _, positions in clusters.groupby(clusters).indices.items():
        positions = np.asarray(list(positions))
        cluster_scores.append(X[positions].T @ (y[positions] - mu[positions]))
    meat = sum(np.outer(s, s) for s in cluster_scores)
    g, n, p = len(cluster_scores), len(y), X.shape[1]
    covariance = bread @ meat @ bread * (g / (g - 1)) * ((n - 1) / (n - p))
    se = np.sqrt(np.diag(covariance))
    return beta, se, covariance, g, n


def odds_ratio_table(beta: np.ndarray, se: np.ndarray, columns: list[str]) -> pd.DataFrame:
    """Term-by-term odds ratios, 95% CI, and p-value."""
    return pd.DataFrame({
        "term": columns,
        "odds_ratio": np.exp(beta),
        "ci_low": np.exp(beta - 1.96 * se),
        "ci_high": np.exp(beta + 1.96 * se),
        "p_value": [normal_p_value(b / s) for b, s in zip(beta, se)],
    })

File: regression.js
// A JavaScript port of regression.py's fit_clustered_logit: logistic MLE via
// IRLS (iteratively reweighted least squares), with a cluster-robust
// sandwich covariance. Written to match the Python implementation term for
// term, including its exact backtracking and finite-sample correction, so
// its output can be checked against the already-trusted Python baseline
// rather than trusted on its own.

function matVec(X, v) {
  // X: array of row arrays, v: array. Returns X @ v.
  return X.map(row => row.reduce((sum, x, j) => sum + x * v[j], 0));
}

function matTVec(X, v) {
  // X^T @ v, where X is n x p and v is length n. Returns length p.
  const p = X[0].length;
  const out = new Array(p).fill(0);
  for (let i = 0; i < X.length; i++) {
    const xi = X[i], vi = v[i];
    for (let j = 0; j < p; j++) out[j] += xi[j] * vi;
  }
  return out;
}

function xtWx(X, w) {
  // X^T @ (X * w[:, None]) -- weighted Gram matrix, p x p.
  const n = X.length, p = X[0].length;
  const out = Array.from({ length: p }, () => new Array(p).fill(0));
  for (let i = 0; i < n; i++) {
    const xi = X[i], wi = w[i];
    for (let a = 0; a < p; a++) {
      const xia_wi = xi[a] * wi;
      for (let b = 0; b < p; b++) out[a][b] += xia_wi * xi[b];
    }
  }
  return out;
}

function solve(A, b) {
  // Solves A @ x = b via Gaussian elimination with partial pivoting.
  // A: p x p, b: length p. A is not mutated.
  const p = A.length;
  const M = A.map((row, i) => [...row, b[i]]);
  for (let col = 0; col < p; col++) {
    let pivotRow = col;
    for (let r = col + 1; r < p; r++) {
      if (Math.abs(M[r][col]) > Math.abs(M[pivotRow][col])) pivotRow = r;
    }
    [M[col], M[pivotRow]] = [M[pivotRow], M[col]];
    const pivot = M[col][col];
    for (let r = 0; r < p; r++) {
      if (r === col) continue;
      const factor = M[r][col] / pivot;
      for (let c = col; c <= p; c++) M[r][c] -= factor * M[col][c];
    }
  }
  return M.map((row, i) => row[p] / row[i]);
}

function inverse(A) {
  // A^-1 via solving A @ x_j = e_j for each standard basis vector.
  const p = A.length;
  const cols = [];
  for (let j = 0; j < p; j++) {
    const e = new Array(p).fill(0);
    e[j] = 1;
    cols.push(solve(A, e));
  }
  // cols[j] is column j of the inverse; transpose to row-major.
  const inv = Array.from({ length: p }, () => new Array(p).fill(0));
  for (let i = 0; i < p; i++) for (let j = 0; j < p; j++) inv[i][j] = cols[j][i];
  return inv;
}

function logLikelihood(X, y, beta) {
  let total = 0;
  for (let i = 0; i < X.length; i++) {
    const linear = Math.max(-30, Math.min(30, matVec([X[i]], beta)[0]));
    // logaddexp(0, linear) = log(1 + exp(linear)), computed stably.
    const logaddexp0 = linear > 0 ? linear + Math.log1p(Math.exp(-linear)) : Math.log1p(Math.exp(linear));
    total += y[i] * linear - logaddexp0;
  }
  return total;
}

function fitClusteredLogit(X, y, clusters) {
  const n = X.length, p = X[0].length;
  let beta = new Array(p).fill(0);
  let currentLL = logLikelihood(X, y, beta);
  let converged = false;

  for (let iter = 0; iter < 100; iter++) {
    const eta = X.map(row => Math.max(-30, Math.min(30, matVec([row], beta)[0])));
    const mu = eta.map(e => 1 / (1 + Math.exp(-e)));
    const w = mu.map(m => Math.max(1e-9, m * (1 - m)));
    const hessian = xtWx(X, w);
    const residual = y.map((yi, i) => yi - mu[i]);
    const score = matTVec(X, residual);
    const step = solve(hessian, score);

    let multiplier = 1.0;
    let betaNext = beta.map((b, j) => b + multiplier * step[j]);
    while (logLikelihood(X, y, betaNext) < currentLL && multiplier > 1e-6) {
      multiplier /= 2;
      betaNext = beta.map((b, j) => b + multiplier * step[j]);
    }
    const maxStep = Math.max(...step.map((s, j) => Math.abs(multiplier * s)));
    beta = betaNext;
    if (maxStep < 1e-9) { converged = true; break; }
    currentLL = logLikelihood(X, y, beta);
  }
  if (!converged) throw new Error("IRLS did not converge");

  const etaFinal = X.map(row => Math.max(-30, Math.min(30, matVec([row], beta)[0])));
  const muFinal = etaFinal.map(e => 1 / (1 + Math.exp(-e)));
  const wFinal = muFinal.map(m => m * (1 - m));
  const bread = inverse(xtWx(X, wFinal));

  const groups = new Map();
  clusters.forEach((c, i) => {
    if (!groups.has(c)) groups.set(c, []);
    groups.get(c).push(i);
  });
  const meat = Array.from({ length: p }, () => new Array(p).fill(0));
  for (const positions of groups.values()) {
    const scoreVec = new Array(p).fill(0);
    for (const i of positions) {
      const resid = y[i] - muFinal[i];
      for (let j = 0; j < p; j++) scoreVec[j] += X[i][j] * resid;
    }
    for (let a = 0; a < p; a++) for (let b = 0; b < p; b++) meat[a][b] += scoreVec[a] * scoreVec[b];
  }

  const g = groups.size;
  // covariance = bread @ meat @ bread * (g/(g-1)) * ((n-1)/(n-p))
  function matMul(A, B) {
    const rA = A.length, cA = A[0].length, cB = B[0].length;
    const out = Array.from({ length: rA }, () => new Array(cB).fill(0));
    for (let i = 0; i < rA; i++) for (let k = 0; k < cA; k++) {
      const aik = A[i][k];
      if (aik === 0) continue;
      for (let j = 0; j < cB; j++) out[i][j] += aik * B[k][j];
    }
    return out;
  }
  const scale = (g / (g - 1)) * ((n - 1) / (n - p));
  const bm = matMul(bread, meat);
  const covRaw = matMul(bm, bread);
  const covariance = covRaw.map(row => row.map(v => v * scale));
  const se = covariance.map((row, i) => Math.sqrt(row[i]));

  return { beta, se, covariance, g, n };
}

function normalPValue(z) {
  // Two-sided p-value for a standard-normal z, via the erf-based normal CDF.
  function erf(x) {
    // Abramowitz & Stegun 7.1.26 approximation, ~1.5e-7 max error.
    const sign = x < 0 ? -1 : 1;
    x = Math.abs(x);
    const a1 = 0.254829592, a2 = -0.284496736, a3 = 1.421413741;
    const a4 = -1.453152027, a5 = 1.061405429, pp = 0.3275911;
    const t = 1 / (1 + pp * x);
    const y = 1 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * Math.exp(-x * x);
    return sign * y;
  }
  return 2 * (1 - 0.5 * (1 + erf(Math.abs(z) / Math.sqrt(2))));
}

function oddsRatioTable(beta, se, columns) {
  return columns.map((term, i) => ({
    term,
    odds_ratio: Math.exp(beta[i]),
    ci_low: Math.exp(beta[i] - 1.96 * se[i]),
    ci_high: Math.exp(beta[i] + 1.96 * se[i]),
    p_value: normalPValue(beta[i] / se[i]),
  }));
}

if (typeof module !== "undefined") {
  module.exports = { fitClusteredLogit, oddsRatioTable, normalPValue };
}
